src/Form/AuthenticationForm.php

```diff
@@ -372,26 +392,187 @@ public function validateForm(array &$form, FormStateInterface $form_state) {
-    }
     catch (\Exception $exception) {
       watchdog_exception('apigee_edge', $exception);
-      if ($this->moduleHandler->moduleExists('dblog')) {
-        $form_state->setError($form, $this->t('Connection failed. Response from Apigee Edge: %response<br>For further information please check the <a href=":url">log messages</a>.', [
-          '%response' => $exception->getMessage(),
-          ':url' => Url::fromRoute('dblog.overview')->toString(),
-        ]));
+
+      $form_state->setError($form, $this->t('@suggestion Error message: %response.', [
+        // Use Markup instead of primitive string to avoid double
+        // sanitizing.
+        '@suggestion' => Markup::create($this->createSuggestion($exception, $key)),
+        '%response' => $exception->getMessage(),
+      ]));
+
+      // Display debug information.
+      $form['debug']['#access'] = TRUE;
+      $form['debug']['debug_text']['#access'] = TRUE;
+      $form['debug']['debug_text']['#value'] = $this->createDebugText($exception, $key, $key_token);
+    }
+  }
+
+  /**
+   * Creates a suggestion text to be displayed in the connection failed message.
+   *
+   * @param \Exception $exception
+   *   The thrown exception during form validation.
+   * @param \Drupal\key\KeyInterface $key
+   *   The used key during form validation.
+   *
+   * @return string
+   *   The suggestion text to be displayed.
+   */
+  protected function createSuggestion(\Exception $exception, KeyInterface $key): string {
+    /** @var \Drupal\apigee_edge\Plugin\KeyType\BasicAuthKeyType $key_type */
+    $key_type = $key->getKeyType();
+
+    // Failed to connect to the Oauth authorization server.
+    if ($exception instanceof OauthAuthenticationException) {
+      /** @var \Drupal\apigee_edge\Plugin\KeyType\OauthKeyType $key_type */
+      $fail_text = $this->t('Failed to connect to the OAuth authorization server.');
+      // General error message.
+      $suggestion = $this->t('@fail_text Check the debug information below for more details.', [
+        '@fail_text' => $fail_text,
+      ]);
+      // Invalid credentials.
+      if ($exception->getCode() === 401) {
+        // Invalid credentials using defined client_id/client_secret.
+        if ($key_type->getClientId($key) === NULL || $key_type->getClientSecret($key) === NULL) {
+          $suggestion = $this->t('@fail_text The given username (%username) or password or client ID (%client_id) or client secret is probably incorrect.', [
+            '@fail_text' => $fail_text,
+            '%client_id' => $key_type->getClientId($key) ?? Oauth::DEFAULT_CLIENT_ID,
+            '%username' => $key_type->getUsername($key),
+          ]);
+        }
+        // Invalid credentials using default client_id/client_secret.
+        else {
+          $suggestion = $this->t('@fail_text The given username (%username) or password is probably incorrect.', [
+            '@fail_text' => $fail_text,
```

Comment:
What if you would inject fail_text with `'!fail_text' => $fail_text->render()`? Using `Markup::create()` would be still necessary above?